File: apps/web/src/lib/plugins/wasiRunner.ts
/**
 * Browser WASI runner — executes arbitrary WASI modules using @bjorn3/browser_wasi_shim.
 *
 * This is the browser counterpart to `diaryx_extism::wasi_runner` on native.
 * Guest plugins call `host_run_wasi_module` with a storage key, arguments,
 * optional stdin, virtual filesystem files, and output file paths.
 * The host loads the WASM bytes from plugin storage, runs the module,
 * and captures stdout/stderr/output files.
 */

// Re-export the @bjorn3/browser_wasi_shim types we use
import {
  File,
  OpenFile,
  PreopenDirectory,
  WASI,
} from '@bjorn3/browser_wasi_shim';

export interface WasiRunRequest {
  module_key: string;
  args: string[];
  stdin?: string;                          // base64
  files?: Record<string, string>;          // path → base64
  output_files?: string[];
}

export interface WasiRunResult {
  exit_code: number;
  stdout: string;                          // base64
  stderr: string;                          // UTF-8 text
  files?: Record<string, string>;          // path → base64
}

/**
 * Run a WASI module in the browser.
 *
 * @param wasmBytes - The raw WASM binary
 * @param args - CLI arguments (argv[0] should be the program name)
 * @param stdinBytes - Optional stdin data
 * @param inputFiles - Virtual filesystem files (path → data)
 * @param outputFilePaths - Paths of output files to capture
 */
export async function runWasiModule(
  wasmBytes: ArrayBuffer,
  args: string[],
  stdinBytes?: Uint8Array,
  inputFiles?: Map<string, Uint8Array>,
  outputFilePaths?: string[],
): Promise<WasiRunResult> {
  // Build the virtual filesystem
  const fileSystem = new Map<string, File>();

  // stdin
  const stdinFile = new File(stdinBytes ?? new Uint8Array(), { readonly: true });

  // stdout/stderr capture
  let stdoutData = new Uint8Array();
  let stderrData = new Uint8Array();

  const stdoutFile = new File(new Uint8Array(), { readonly: false });
  const stderrFile = new File(new Uint8Array(), { readonly: false });

  // Add input files to the virtual FS
  if (inputFiles) {
    for (const [path, data] of inputFiles) {
      fileSystem.set(path.replace(/^\//, ''), new File(data, { readonly: true }));
    }
  }

  // Add placeholder output files
  if (outputFilePaths) {
    for (const path of outputFilePaths) {
      const cleanPath = path.replace(/^\//, '');
      if (!fileSystem.has(cleanPath)) {
        fileSystem.set(cleanPath, new File(new Uint8Array(), { readonly: false }));
      }
    }
  }

  // Set up WASI file descriptors:
  // fd 0 = stdin, fd 1 = stdout, fd 2 = stderr, fd 3 = preopened "/"
  const fds = [
    new OpenFile(stdinFile),
    new OpenFile(stdoutFile),
    new OpenFile(stderrFile),
    new PreopenDirectory('/', fileSystem),
  ];

  const fullArgs = ['program', ...args];
  const wasi = new WASI(fullArgs, [], fds, { debug: false });

  // Compile and instantiate
  const module = await WebAssembly.compile(wasmBytes);
  const instance = await WebAssembly.instantiate(module, {
    wasi_snapshot_preview1: wasi.wasiImport,
  });

  // Run the module
  let exitCode = 0;
  try {
    wasi.initialize(instance as any);
    // Call _start if it exists
    const start = (instance.exports as any)._start;
    if (typeof start === 'function') {
      start();
    }
  } catch (e: any) {
    // WASI proc_exit throws with the exit code
    if (e instanceof Error && e.message?.includes('exit')) {
      // Try to extract exit code
      const match = e.message.match(/exit.*?(\d+)/i);
      exitCode = match ? parseInt(match[1], 10) : 1;
    } else if (typeof e === 'object' && 'code' in e) {
      exitCode = (e as any).code;
    } else {
      throw e;
    }
  }

  // Capture stdout
  stdoutData = new Uint8Array(stdoutFile.data);

  // Capture stderr
  stderrData = new Uint8Array(stderrFile.data);

  // Base64 encode stdout
  const stdoutB64 = uint8ArrayToBase64(stdoutData);

  // Decode stderr as UTF-8 text
  const stderrText = new TextDecoder().decode(stderrData);

  // Capture output files
  let capturedFiles: Record<string, string> | undefined;
  if (outputFilePaths && outputFilePaths.length > 0) {
    capturedFiles = {};
    for (const path of outputFilePaths) {
      const cleanPath = path.replace(/^\//, '');
      const file = fileSystem.get(cleanPath);
      if (file && file.data && file.data.length > 0) {
        capturedFiles[path] = uint8ArrayToBase64(file.data);
      }
    }
    if (Object.keys(capturedFiles).length === 0) {
      capturedFiles = undefined;
    }
  }

  return {
    exit_code: exitCode,
    stdout: stdoutB64,
    stderr: stderrText,
    files: capturedFiles,
  };
}

/** Convert a Uint8Array to a base64 string. */
function uint8ArrayToBase64(bytes: Uint8Array): string {
  let binary = '';
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
}

/** Convert a base64 string to a Uint8Array. */
function base64ToUint8Array(b64: string): Uint8Array {
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

/**
 * Handle a `host_run_wasi_module` call from an Extism guest plugin.
 *
 * Loads WASM bytes from localStorage (plugin storage), decodes base64 inputs,
 * runs the WASI module, and returns the result.
 */
export async function handleHostRunWasiModule(
  request: WasiRunRequest,
): Promise<WasiRunResult> {
  // Load WASM bytes from plugin storage (localStorage)
  const storageKey = `diaryx-plugin:${request.module_key}`;
  const raw = localStorage.getItem(storageKey);
  if (!raw) {
    return {
      exit_code: -1,
      stdout: '',
      stderr: `Module not found in storage: ${request.module_key}`,
    };
  }

  // Parse the stored data (same format as host_storage_get: {data: base64})
  let wasmB64: string;
  try {
    const parsed = JSON.parse(raw);
    wasmB64 = parsed.data;
  } catch {
    return {
      exit_code: -1,
      stdout: '',
      stderr: `Invalid storage format for module: ${request.module_key}`,
    };
  }

  const wasmBytes = base64ToUint8Array(wasmB64).buffer as ArrayBuffer;

  // Decode stdin
  const stdinBytes = request.stdin ? base64ToUint8Array(request.stdin) : undefined;

  // Decode input files
  let inputFiles: Map<string, Uint8Array> | undefined;
  if (request.files) {
    inputFiles = new Map();
    for (const [path, b64] of Object.entries(request.files)) {
      inputFiles.set(path, base64ToUint8Array(b64));
    }
  }

  return runWasiModule(
    wasmBytes,
    request.args,
    stdinBytes,
    inputFiles,
    request.output_files,
  );
}

File: crates/diaryx_extism/src/wasi_runner.rs
//! Standalone WASI module runner.
//!
//! Allows Extism guest plugins to execute arbitrary WASI programs (e.g. pandoc,
//! typst) by delegating to the host runtime. The host loads the WASI module
//! from plugin storage, sets up a sandboxed filesystem, and captures
//! stdout/stderr.
//!
//! This module is gated behind the `wasi-runner` feature.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};

/// Request to run a WASI module.
#[derive(Debug, Deserialize)]
pub struct WasiRunRequest {
    /// Storage key holding the WASM bytes (retrieved via `PluginStorage::get`).
    pub module_key: String,
    /// CLI arguments (argv) passed to the WASI program.
    pub args: Vec<String>,
    /// Optional stdin data (base64-encoded).
    #[serde(default)]
    pub stdin: Option<String>,
    /// Virtual filesystem files to make available (path → base64 content).
    #[serde(default)]
    pub files: Option<HashMap<String, String>>,
    /// Paths of output files to capture after execution (path → base64 content).
    #[serde(default)]
    pub output_files: Option<Vec<String>>,
}

/// Result of running a WASI module.
#[derive(Debug, Serialize)]
pub struct WasiRunResult {
    /// Process exit code (0 = success).
    pub exit_code: i32,
    /// Captured stdout (base64-encoded).
    pub stdout: String,
    /// Captured stderr (UTF-8 text).
    pub stderr: String,
    /// Captured output files (path → base64 content).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub files: Option<HashMap<String, String>>,
}

/// Run a WASI module with the given bytes, arguments, and virtual filesystem.
///
/// This creates a temporary directory, writes input files, runs the module's
/// `_start` export, captures stdout/stderr, and reads back output files.
pub fn run_wasi_module(
    wasm_bytes: &[u8],
    args: &[String],
    stdin: Option<&[u8]>,
    files: Option<&HashMap<String, Vec<u8>>>,
    output_files: Option<&[String]>,
) -> Result<WasiRunResult, String> {
    use base64::Engine;
    use wasi_common::pipe::{ReadPipe, WritePipe};
    use wasi_common::sync::{Dir, WasiCtxBuilder, ambient_authority};
    use wasmtime::{Engine as WasmEngine, Linker, Module, Store};

    // Create a temp directory for the virtual filesystem
    let temp_dir =
        tempfile::tempdir().map_err(|e| format!("Failed to create temp directory: {e}"))?;
    let work_dir = temp_dir.path();

    // Write input files to the temp directory
    if let Some(input_files) = files {
        for (path, data) in input_files {
            let full_path = work_dir.join(path.trim_start_matches('/'));
            if let Some(parent) = full_path.parent() {
                std::fs::create_dir_all(parent)
                    .map_err(|e| format!("Failed to create dir {}: {e}", parent.display()))?;
            }
            std::fs::write(&full_path, data)
                .map_err(|e| format!("Failed to write {}: {e}", full_path.display()))?;
        }
    }

    // Build WASI context
    let mut ctx_builder = WasiCtxBuilder::new();

    // Set arguments: argv[0] is the program name, then user args
    let mut full_args: Vec<String> = vec!["program".to_string()];
    full_args.extend_from_slice(args);
    ctx_builder
        .args(&full_args)
        .map_err(|e| format!("Failed to set args: {e}"))?;

    // Set stdin
    if let Some(stdin_data) = stdin {
        ctx_builder.stdin(Box::new(ReadPipe::from(stdin_data.to_vec())));
    }

    // Capture stdout/stderr
    let stdout_pipe = WritePipe::new_in_memory();
    let stderr_pipe = WritePipe::new_in_memory();
    ctx_builder.stdout(Box::new(stdout_pipe.clone()));
    ctx_builder.stderr(Box::new(stderr_pipe.clone()));

    // Preopen the work directory as "." — this is the only directory the WASI
    // module can access. We do NOT preopen "/" to prevent escape from the sandbox.
    let preopen_dir = Dir::open_ambient_dir(work_dir, ambient_authority())
        .map_err(|e| format!("Failed to open work dir: {e}"))?;
    ctx_builder
        .preopened_dir(preopen_dir, ".")
        .map_err(|e| format!("Failed to preopen dir: {e}"))?;

    let wasi_ctx = ctx_builder.build();

    // Compile and instantiate
    let engine = if let Some(config) = crate::platform_wasmtime_config() {
        WasmEngine::new(&config).map_err(|e| format!("Failed to create wasmtime engine: {e}"))?
    } else {
        WasmEngine::default()
    };
    let module = Module::new(&engine, wasm_bytes)
        .map_err(|e| format!("Failed to compile WASI module: {e}"))?;

    let mut linker = Linker::new(&engine);
    wasi_common::sync::add_to_linker(&mut linker, |s| s)
        .map_err(|e| format!("Failed to add WASI to linker: {e}"))?;

    let mut store = Store::new(&engine, wasi_ctx);
    let instance = linker
        .instantiate(&mut store, &module)
        .map_err(|e| format!("Failed to instantiate WASI module: {e}"))?;

    // Call _start
    let start_fn = instance
        .get_typed_func::<(), ()>(&mut store, "_start")
        .map_err(|e| format!("WASI module has no _start export: {e}"))?;

    let exit_code = match start_fn.call(&mut store, ()) {
        Ok(()) => 0,
        Err(e) => {
            // Check for WASI proc_exit (normal termination)
            if let Some(exit) = e.downcast_ref::<wasi_common::I32Exit>() {
                exit.0
            } else {
                // Extract stdout/stderr even on error for diagnostics
                let stderr_bytes = stderr_pipe
                    .try_into_inner()
                    .map(|c| c.into_inner())
                    .unwrap_or_default();
                let stderr_text = String::from_utf8_lossy(&stderr_bytes);
                return Err(format!("WASI module trapped: {e}\nstderr: {stderr_text}"));
            }
        }
    };

    // Drop the store so pipes can be consumed
    drop(store);

    // Read captured stdout
    let stdout_bytes = stdout_pipe
        .try_into_inner()
        .map(|c| c.into_inner())
        .unwrap_or_default();
    let stdout_b64 = base64::engine::general_purpose::STANDARD.encode(&stdout_bytes);

    // Read captured stderr
    let stderr_bytes = stderr_pipe
        .try_into_inner()
        .map(|c| c.into_inner())
        .unwrap_or_default();
    let stderr_text = String::from_utf8_lossy(&stderr_bytes).to_string();

    // Read output files
    let captured_files = if let Some(paths) = output_files {
        let mut result = HashMap::new();
        for path in paths {
            let full_path = work_dir.join(path.trim_start_matches('/'));
            if full_path.exists() {
                match std::fs::read(&full_path) {
                    Ok(data) => {
                        result.insert(
                            path.clone(),
                            base64::engine::general_purpose::STANDARD.encode(&data),
                        );
                    }
                    Err(e) => {
                        log::warn!(
                            "[wasi_runner] Failed to read output file {}: {e}",
                            full_path.display()
                        );
                    }
                }
            }
        }
        if result.is_empty() {
            None
        } else {
            Some(result)
        }
    } else {
        None
    };

    Ok(WasiRunResult {
        exit_code,
        stdout: stdout_b64,
        stderr: stderr_text,
        files: captured_files,
    })
}
